src/test.rs

```diff
@@ -192,15 +199,95 @@ fn it_works_decode_with_version_0() {
+    };
+    let mut constraints: Vec<u8> = Vec::new();
+    contract.encode(&mut constraints);
+
+    let method = Method {
+        name: "method_test".to_string(),
+        block_cooldown: None,
+        constraints: Some(constraints),
+    };
+    let mut methods: Vec<(String, Method)> = Default::default();
+    methods.push((method.name.clone(), method.clone()));
+
+    let module = Module {
+        name: "module_test".to_string(),
+        block_cooldown: None,
+        methods,
+    };
+
+    let mut modules: Vec<(String, Module)> = Default::default();
+    modules.push((module.name.clone(), module.clone()));
+
+    let cennznut = CENNZnutV0 { modules };
+
+    assert_eq!(
+        cennznut.encode(),
+        vec![
+            0, 0, 0, 64, 109, 111, 100, 117, 108, 101, 95, 116, 101, 115, 116, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 74, 109, 101, 116, 104, 111, 100, 95, 116,
+            101, 115, 116, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 192,
+            128, 16, 246, 0, 0, 0, 0, 0, 0, 0, 128, 16, 178, 128, 0, 0, 0, 0, 0, 0, 0, 224, 116,
+            101, 115, 116, 105, 110, 103, 5, 0, 0, 1, 0, 5, 0, 1, 1, 1,
+        ]
+    );
+}
+
+#[test]
+fn it_works_decode_with_constraints() {
+    let encoded_cennznut: Vec<u8> = vec![
+        0, 0, 0, 64, 109, 111, 100, 117, 108, 101, 95, 116, 101, 115, 116, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 74, 109, 101, 116, 104, 111, 100, 95, 116, 101, 115,
+        116, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 192, 128, 16, 246,
+        0, 0, 0, 0, 0, 0, 0, 128, 16, 178, 128, 0, 0, 0, 0, 0, 0, 0, 224, 116, 101, 115, 116, 105,
+        110, 103, 5, 0, 0, 1, 0, 5, 0, 1, 1, 1,
+    ];
+    CENNZnutV0::decode(&mut &encoded_cennznut[..]).unwrap();
+}
+
+#[test]
+#[should_panic(expected = "invalid constraints codec")]
+fn it_works_decode_with_valid_constraints() {
+    let encoded_cennznut: Vec<u8> = vec![
+        0, 0, 0, 64, 109, 111, 100, 117, 108, 101, 95, 116, 101, 115, 116, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 109, 101, 116, 104, 111, 100, 95, 116, 101, 115,
+        116, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    ];
+
+    let bad_type_id: Vec<u8> = vec![0, 0b1000_0000, 0b0000_0001, 0b0000_0001];
+    let n_too_short: Vec<u8> = vec![0, 1];
+    let n_too_large: Vec<u8> = vec![0, 0b1000_0000, 0b1000_0000, 0b0000_1111];
+
+    let encoded_with_bad_type_id: Vec<u8> = [encoded_cennznut.clone(), bad_type_id].concat();
+    let encoded_with_n_too_short: Vec<u8> = [encoded_cennznut.clone(), n_too_short].concat();
+    let encoded_with_n_too_large: Vec<u8> = [encoded_cennznut.clone(), n_too_large].concat();
+
+    CENNZnutV0::decode(&mut &encoded_with_bad_type_id[..]).unwrap();
+    CENNZnutV0::decode(&mut &encoded_with_n_too_short[..]).unwrap();
```

Comment:
Do the lines after the first failure (`.unwrap()`) actually execute or does this exit after the first failure?